Add Next-use distance analysis

kmitropoulou opened the pull request:
This analysis is used by ERS (Early Register Spilling) to find which register to spill. 

It iterates among the uses of a definition. It calculates the shortest distance between the definition and the use. The next-use distance of a register is the shortest distance among all the uses. 

There is a penalty for the values that are live across loops.

ERS chooses to spill the register with the longest next-use distance.

Commit:
Add Next-use distance analysis

ronlieb commented:
cc @linuxrocks123

ruiling commented on llvm/lib/Target/AMDGPU/AMDGPUNextUseAnalysis.cpp:
```diff
@@ -0,0 +1,573 @@
+}
+
+std::optional<uint64_t>
+AMDGPUNextUseAnalysis::getNextUseDistance(Register DefReg, MachineInstr *CurMI,
```
Now that we have `LiveIntervals` in the `ERS`. You can pass in here, which can help you calculate the distance.

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -1602,6 +1607,9 @@ void GCNPassConfig::addOptimizedRegAlloc() {
     insertPass(&LiveVariablesID, &SIOptimizeVGPRLiveRangeLegacyID);
 
+  if (OptNextUseAnalysis)
+    insertPass(&LiveVariablesID, &AMDGPUNextUseAnalysisID);
```
Why? Shouldn't this only be added when it's requested by another pass?

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUNextUseAnalysis.h:
```diff
@@ -0,0 +1,173 @@
+  /// them and fills in \p ShortestPathTable.
+  void calculateShortestPaths(MachineFunction &MF);
+  /// If the path from \p MI to \p UseMI does not cross any loops, then this
+  /// \returns the shortest instruction distance between them.
```
Should also describe what it returns if there are loops.

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUNextUseAnalysis.h:
```diff
@@ -0,0 +1,173 @@
+  /// Given \p CurMI in a loop and \p UseMI outside the loop, this function
+  /// returns the minimum instruction path between \p CurMI and \p UseMI.
+  /// Please note that since \p CurMI is in a loop we don't care about the
+  /// exact position of the instruction in the block because we making a rough
```
```suggestion
  /// exact position of the instruction in the block because we're making a rough
```

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUNextUseAnalysis.cpp:
```diff
@@ -0,0 +1,573 @@
+  Loop1->getLoopLatches(Latches);
+  auto It = llvm::find(Latches, From);
+  MachineBasicBlock *LoopHeader = Loop1->getHeader();
+  return It != Latches.end() && From->isSuccessor(To) && To == LoopHeader;
```
You can check `isSuccessor` right at the top of the function and exit early if it isn't. Likewise, you can check if To is Loop1's header before getting the loop latches and calling find on them.

Commit:
Update next-use distance for uses inside a loop nest with multiple loops at each
loop level.

Commit:
Address review comments

Commit:
Update the loop weight to be 1000 to the power of the loop depth

Currently, the loop weight is the loop depth multiplied by 1000(which represents
the loop iterations).
Now, the loop weight is 1000 to the power of the loop depth.

Commit:
Remove Path from ShortestPathTable

Commit:
Small refactoring

Commit:
Rename tests

Commit:
Calculate next-use distance on demand

ronlieb reviewed (changes requested):
Please hold off on landing this PR as we are looking at an alternative solution to cover this one and https://github.com/llvm/llvm-project/pull/156079
